src/plugins/ui_actions/public/context_menu/build_eui_context_menu_panels.tsx

```diff
@@ -41,138 +44,178 @@ interface ActionWithContext<Context extends BaseContext = BaseContext> {
-    actions,
-    closeMenu,
-  });
-
-  return {
-    id: 'mainMenu',
-    title,
-    items: menuItems,
-  };
 }
 
 /**
- * Transform an array of Actions into the shape needed to build an EUIContextMenu
+ * Transforms an array of Actions to the shape EuiContextMenuPanel expects.
  */
-async function buildEuiContextMenuPanelItems({
+export async function buildContextMenuForActions({
   actions,
-  closeMenu,
-}: {
-  actions: ActionWithContext[];
-  closeMenu: () => void;
-}) {
-  actions = sortBy(
-    actions,
-    (a) => -1 * (a.action.order ?? 0),
-    (a) => a.action.type,
-    (a) => a.action.getDisplayName({ ...a.context, trigger: a.trigger })
-  );
-
-  const items: EuiContextMenuPanelItemDescriptor[] = new Array(actions.length);
-  const promises = actions.map(async ({ action, context, trigger }, index) => {
-    const isCompatible = await action.isCompatible({
-      ...context,
-      trigger,
-    });
-    if (!isCompatible) {
-      return;
+  title = defaultTitle,
+  closeMenu = () => {},
+}: BuildContextMenuParams): Promise<EuiContextMenuPanelDescriptor[]> {
+  const panels: Record<string, PanelDescriptor> = {
+    mainMenu: {
+      id: 'mainMenu',
+      title,
+      items: [],
+    },
+  };
+  const promises = actions.map(async (item) => {
+    const { action } = item;
+    const context: ActionExecutionContext<object> = { ...item.context, trigger: item.trigger };
+    const isCompatible = await item.action.isCompatible(context);
+    if (!isCompatible) return;
+    let parentPanel = '';
+    let currentPanel = '';
+    if (action.grouping && action.grouping) {
+      for (let i = 0; i < action.grouping.length; i++) {
+        const group = action.grouping[i];
+        currentPanel = group.id;
+        if (!panels[currentPanel]) {
+          const name = group.getDisplayName ? group.getDisplayName(context) : group.id;
+          panels[currentPanel] = {
+            id: currentPanel,
+            title: name,
+            items: [],
+            _level: i,
+            _icon: group.getIconType ? group.getIconType(context) : 'empty',
+          };
+          if (parentPanel) {
+            panels[parentPanel].items!.push({
+              name,
+              panel: currentPanel,
+              icon: group.getIconType ? group.getIconType(context) : 'empty',
+              _order: group.order || 0,
+              _title: group.getDisplayName ? group.getDisplayName(context) : '',
+            });
+          }
+        }
+        parentPanel = currentPanel;
+      }
     }
-
-    items[index] = await convertPanelActionToContextMenuItem({
-      action,
-      actionContext: context,
-      trigger,
-      closeMenu,
+    panels[parentPanel || 'mainMenu'].items!.push({
+      name: action.MenuItem
+        ? React.createElement(uiToReactComponent(action.MenuItem), { context })
+        : action.getDisplayName(context),
+      icon: action.getIconType(context),
+      panel: _.get(action, 'childContextMenuPanel.id'),
```

Comment:
is this `panel: _.get(action, 'childContextMenuPanel.id')` used? I guess this is an artefact from older implementation, should we remove it?

Reply:
Removed.